timetracker/api/views.py

```diff
@@ -1,22 +1,139 @@
 from rest_framework.viewsets import ModelViewSet
+from rest_framework.response import Response
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.decorators import api_view
 from timetracker.models import Employee
 from timetracker.models import Checking
-from timetracker.models import Vacation
 from .serializers import EmployeeSerializer
 from .serializers import CheckingSerializer
 from .serializers import VacationSerializer
 
 
 class EmployeeViewSet(ModelViewSet):
+    permission_classes = [IsAuthenticated]
     queryset = Employee.objects.all()
     serializer_class = EmployeeSerializer
 
+    YEAR_QUARTERS = {
+        '1': [1, 3],
+        '2': [3, 6],
+        '3': [6, 9],
+        '4': [9, 12]
+    }
+
+    @action(methods=['POST'], detail=True, permission_classes=[IsAuthenticated],
+            url_path='set-check', url_name='set_check')
+    def set_checking(self, request, pk=None):
+        check = request.data['check'] if 'check' in request.data else ''
+        checking_serializer = CheckingSerializer(data={'check': check, 'employee': pk})
+        if checking_serializer.is_valid():
+            checking_serializer.save()
+            return Response({'result': checking_serializer.data, 'error': None}, status=status.HTTP_201_CREATED)
+        return Response({'error': json.dumps(checking_serializer.errors)}, status=status.HTTP_400_BAD_REQUEST)
+
+    @action(methods=['POST'], detail=True, permission_classes=[IsAuthenticated],
+            url_path='set-vacation', url_name='set_vacation')
+    def set_vacation(self, request, pk=None):
+        vacation_serializer = VacationSerializer(
+            data={'employee': pk,
+                  'description': request.data['description'] if 'description' in request.data else ''})
+        if vacation_serializer.is_valid():
+            vacation_serializer.save()
+            return Response({'result': vacation_serializer.data, 'error': None}, status=status.HTTP_201_CREATED)
+        return Response({'error': json.dumps(vacation_serializer.errors)}, status=status.HTTP_400_BAD_REQUEST)
+
+    @action(methods=['GET'], detail=True, permission_classes=[IsAuthenticated],
+            url_path='work-avg', url_name='work_avg')
+    def get_work_avg(self, request, pk=None):
+        try:
+            employee = Employee.objects.get(id=pk)
+        except:
+            employee = None
+        if employee is None:
+            return Response({'error': 'Employee is not found for id {}'.format(pk)}, status=status.HTTP_404_NOT_FOUND)
+        check_in_count = employee.checking_set.filter(check='in').count()
+        check_out_count = employee.checking_set.filter(check='out').count()
+        average = (check_in_count + check_out_count) / 2
+        return Response({'result': average, 'error': None}, status=status.HTTP_200_OK)
+
+    '''
+
+        1- Working hours are the checkim times 8 assuming 8 is a day working hours.
```

Comment:
@gsalah18 this is not how checkin/out should be.

Imagine that employees use a finger print machine to checkin/checkout, then when employee puts his finger in the machine, it will identify him, and save a checkin entry with the current datetime, same for checkout.

This means that the number of hours the employee worked that day equals: Checkout datetime - Checkin datetime.